experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,399 @@
+
+	repoLock sync.Mutex
+	repos    []github.Repo
+}
+
+func NewServer(name, creds string, hmac []byte, gc *git.Client, ghc *github.Client, repos []github.Repo) *Server {
+	return &Server{
+		hmacSecret:  hmac,
+		credentials: creds,
+		botName:     name,
+
+		gc:  gc,
+		ghc: ghc,
+		log: logrus.StandardLogger().WithField("client", "cherrypicker"),
+
+		bare:     &http.Client{},
+		patchURL: "https://patch-diff.githubusercontent.com",
+
+		repos: repos,
+	}
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	eventType, eventGUID, payload, ok := hook.ValidateWebhook(w, r, s.hmacSecret)
+	if !ok {
+		return
+	}
+	fmt.Fprint(w, "Event received. Have a nice day.")
+
+	if err := s.handleEvent(eventType, eventGUID, payload); err != nil {
+		logrus.WithError(err).Error("Error parsing event.")
+	}
+}
+
+func (s *Server) handleEvent(eventType, eventGUID string, payload []byte) error {
+	switch eventType {
+	case "issue_comment":
+		var ic github.IssueCommentEvent
+		if err := json.Unmarshal(payload, &ic); err != nil {
+			return err
+		}
+		go func() {
+			if err := s.handleIssueComment(ic); err != nil {
+				s.log.WithError(err).Info("Cherry-pick failed.")
+			}
+		}()
+	case "pull_request":
+		var pr github.PullRequestEvent
+		if err := json.Unmarshal(payload, &pr); err != nil {
+			return err
+		}
+		go func() {
+			if err := s.handlePullRequest(pr); err != nil {
+				s.log.WithError(err).Info("Cherry-pick failed.")
+			}
+		}()
+	default:
+		return fmt.Errorf("received an event of type %q but didn't ask for it", eventType)
+	}
+	return nil
+}
+
+func (s *Server) handleIssueComment(ic github.IssueCommentEvent) error {
+	// Only consider new comments in PRs.
+	if !ic.Issue.IsPullRequest() || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	cherryPickMatches := cherryPickRe.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if len(cherryPickMatches) == 0 || len(cherryPickMatches[0]) != 2 {
+		return nil
+	}
+	targetBranch := cherryPickMatches[0][1]
+
+	if targetBranch == "master" {
+		return s.ghc.CreateComment(org, repo, num, "I don't cherrypick pull requests onto master.")
+	}
+
+	if ic.Issue.State != "closed" {
+		// Only members should be able to do cherry-picks.
+		ok, err := s.ghc.IsMember(org, commentAuthor)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			resp := fmt.Sprintf("Only [%s](https://github.com/orgs/%s/people) org members may request cherry picks. You can still do the cherry-pick manually.", org, org)
+			s.log.Info(resp)
+			return s.ghc.CreateComment(org, repo, num, resp)
+		}
+		resp := fmt.Sprintf("@%s once the present PR merges, I will cherry-pick it on top of %s in a new PR and assign it to you.", commentAuthor, targetBranch)
+		return s.ghc.CreateComment(org, repo, num, resp)
```

Comment:
Please use `plugins.FormatResponseRaw` here and below since these messages are in response to a comment.

Reply:
done